### ISSUE TITLE
RPA | CICD Automation

### ISSUE THREAD

**Issue by JaleelBennett** [JaleelB/work-issues#6]
Here is the JIRA story i am working on:

RPA | CICD Automation

Description

As DevOps engineers, we need to facilitate automated code deployments for our RPA Process, in order to speed time to delivery while maintaining our guard rails.

Issue
RPA is becoming a larger need at BKU. As this need grows, the DevOps team will need to facilitate CICD and DevOps processes to speed time to delivery while maintaining the appropriate checks. This will require a discovery of the current SDLC – including the responsible parties. This will require several technical steps. This specific story is to do the discovery work, to put together a roadmap and plan to determine if this should be handled a operational work or project work.

DoD
Determine if this is a good candidate for our cicd patterns.
Determine the roadmap
Determine the effort required
Determine the approximate timelines
Discovery to provide everything required from which to create initial project stories.
Review [UiPath CLI · Actions · GitHub Marketplace](https://github.com/marketplace/actions/setup-uipath-cli) (Instead of hard linking UIPath CLI URL) Will need review.

 

Potential Starting Points:

 

[Ultimate CI/CD pipelines using UiPath GitHub Actions and PowerShell - RPABOTS.WORLD (rpabotsworld.com)](https://rpabotsworld.com/uipath-github-actions/)

[Ultimate CI/CD pipelines using UiPath GitHub Actions and PowerShell | Part 2 (youtube.com)Ultimate CI/CD pipelines using UiPath GitHub Actions and PowerShell | Part 2 (youtube.com)](https://www.youtube.com/watch?v=TQD8c9gZIVE)

[CI/CD Pipelines for UiPath Process Deployment Series: GitHub Actions (youtube.com)](https://www.youtube.com/watch?v=musSykryH7U)

[andrianoid/helloUiPath-sample (github.com)andrianoid/helloUiPath-sample (github.com)](https://github.com/andrianoid/helloUiPath-sample)

[rpabotsworld/uipath-with-github-action](https://github.com/rpabotsworld/uipath-with-github-action)

[CI/CD With GitHub Actions - Help / Orchestrator - UiPath Community Forum](https://forum.uipath.com/t/ci-cd-with-github-actions/237060/8)

Here are the acceptance criteria:
- Artifact showing Roadmap, effort, and timelines
- Artifact of determination between Project vs operational
- If Determined to be a viable candidate, Epic needs to be created on the Cloud Program board. (We want to move away from using the rolling 2023 cicd epic for example)

----------

Here are some notes I made from the RPA team:
RPA Software Development Lifecycle (SDLC)

Owned by [Steven Luengo](https://bankunited.atlassian.net/wiki/people/5f454d86d0884f0049e91da0?ref=confluence&src=profilecard)

Last updated: [Apr 08, 2024](https://bankunited.atlassian.net/wiki/pages/diffpagesbyversion.action?pageId=3189800982&selectedPageVersions=3&selectedPageVersions=4)2 min read
1 person viewed
[Overview](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3189800982/RPA+Software+Development+Lifecycle+SDLC#Overview)
[Development details](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3189800982/RPA+Software+Development+Lifecycle+SDLC#Development-details)
[Environment Details](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3189800982/RPA+Software+Development+Lifecycle+SDLC#Environment-Details)
 

 

Overview
 

[RPA Software Development Lifecycle Walkthrough-20240404_153705-Meeting Recording.mp4](https://bankunited-my.sharepoint.com/:v:/p/sluengo/ESljRVBjLq5KlFeUy6QjIBgB-wjAb1rOalbAYXuQmmJhlQ?e=ZoTnTF&nav=eyJyZWZlcnJhbEluZm8iOnsicmVmZXJyYWxBcHAiOiJTdHJlYW1XZWJBcHAiLCJyZWZlcnJhbFZpZXciOiJTaGFyZURpYWxvZy1MaW5rIiwicmVmZXJyYWxBcHBQbGF0Zm9ybSI6IldlYiIsInJlZmVycmFsTW9kZSI6InZpZXcifX0%3D)

An initial evaluation is done to see if the requested automation is both feasible and profitable.

If proposed automation passes that initial analysis, a process definition document (PDD) is created which outlines the procedure a human goes through to complete the process.

Essentially a requirements document

Screenshots and Flowcharts are usually provided.

Question: Where are these documents currently stored? 

A solution design document is then created by one of the RPA architect or developer to automate the process based on the PDD. This is usually produced in Visio.

Question: Where are these documents currently stored? 

Once solution design document is approved, development begins.

There are only two environments: UAT and Production.

QA Engineers:

Rakesh

Apeksha

Question:  Are we working in Sprints? Kanban? Releases?

Branching Strategy:

Feature branches are created from UAT.

Development is done and Pull requests are done against UAT.

If approved, merged into UAT 

Testing is then done in UAT.

If testing passes, PR is made into Master and deployed into Master.

 

 

Development details
Repository Created using current Repo automation.

Need to add Scripts\ folder to house required PowerShell scripts to deploy.

Need to update the template.

Master and uat branches

*.xaml is the code representation of the flows

Feature branch naming convention?

Development is either a feature enhancement or incident.

Feature enhancement do not currently have a naming convention, this should be standardized to use Jira story like others.

CLOUDOPS/CPSM prefix.

Incidents are prefixed with INC ticket number.

INC*

Projects are compiled using project.json file.

Example: [bankunited/RPA-CustBankLabStbyPerfStbyPerfPurch_GL_800001_850001: RPA Team Project (github.com)](https://github.com/bankunited/RPA-CustBankLabStbyPerfStbyPerfPurch_GL_800001_850001)

When packages are deployed into their tenants, the developer needs to manually update the version to latest.

This is something we can look into automating if the PowerShell scripts don’t do this already.

Environment Details
UIPath Orchestrator Version: 2022.10.6

API URL : [Swagger UI (bankunited.com)](https://burpauat.bankunited.com/swagger/index.html)

UAT Tenant:

URL: https://burpauat.bankunited.com/

Production Tenant:

URL: https://burpa.bankunited.com/

RPA Continuous Integration and Deployment (CI/CD)

Owned by [Steven Luengo](https://bankunited.atlassian.net/wiki/people/5f454d86d0884f0049e91da0?ref=confluence&src=profilecard)

Last updated: [Apr 08, 2024](https://bankunited.atlassian.net/wiki/pages/diffpagesbyversion.action?pageId=3192357013&selectedPageVersions=1&selectedPageVersions=2)3 min read
1 person viewed
[Overview](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Overview)
[Pipeline Diagram](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Pipeline-Diagram)
[Parent Workflow](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Parent-Workflow)
[Example Pipeline](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Example-Pipeline)
[Environmental Variables](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Environmental-Variables)
[Print Details](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Print-Details)
[Cleanup](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Cleanup)
[Build UIPath Package](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Build-UIPath-Package)
[Deploy UIPath Package](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Deploy-UIPath-Package)
[Verify](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Verify)
[Future State](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Future-State)
[Resources](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192357013/RPA+Continuous+Integration+and+Deployment+CI+CD#Resources)
 

 

Overview
 

UIPath projects are deployed using the project.json files found in each repository.

CI/CD pipelines need to be run on Windows Runners.

Projects need to be updated with Scripts/ folder which will contain PowerShell Scripts to assist in deployment.

UIPath powershell scripts available from UIPath repository

Where are these scripts?

[UiPath DevOps Scripts - RPA Component | UiPath Marketplace | Overview](https://marketplace.uipath.com/listings/uipath-devops-scripts4936)

[UiPath-DevOps-Scripts/docs/UiPathDeploy.md at main · UiPath-Services/UiPath-DevOps-Scripts (github.com)](https://github.com/UiPath-Services/UiPath-DevOps-Scripts/blob/main/docs/UiPathDeploy.md)

UiPathDeploy.ps1

UiPathPack.ps1

Based on their branching strategy, Feature > UAT > Master:

Feature branches should just attempt to “package/build” the project and run any security or codacy scan (if possible).

Pull Requests should trigger a deployment to UAT tenant.

If a deployment to UAT is successful (use status checks here), then it should require at least 1 approval to merge into master.

When merged into master, trigger production deploy to Prod Tenant

Verify you can do this with RPA team or if they’d rather have it work like Mulesoft, where it will require an additional gate to release to production after merging to master.

We will require a service account of some kind in order to get access to parameters required for this automation:

Client ID

User Key

Etc.

 

 

Pipeline Diagram

TBD (Lucid Chart link here)

Parent Workflow
TBD (Link to GitHub Parent Workflow Repository)

 

Example Pipeline
Environmental Variables
 

ORCH_FOLDER: Folder within Orchestrator where project is (need to be provided by developer)

Open image-20240408-142856.png
image-20240408-142856.png
 

ORCH_URL: 

ORCH_TENANT:

ORCH_CLIENT_ID:

ORCH_USER_KEY:

ORCH_ACC_NAME (Organization ID):

All of these are found in Tenant > Profile in upper right > Preferences > Privacy & Security > View API Access (All values here)

CLI_URL:

We need the path to download the UIPath CLI Download Location.

Open image-20240408-142935.png
image-20240408-142935.png
 

Print Details
Open image-20240408-142956.png
image-20240408-142956.png
 

 

Cleanup
Open image-20240408-143023.png
image-20240408-143023.png
 

Build UIPath Package
We use the UiPathPack.ps1 script to build the project.json file.

[UiPath-DevOps-Scripts/docs/UiPathPack.md at main · UiPath-Services/UiPath-DevOps-Scripts (github.com)](https://github.com/UiPath-Services/UiPath-DevOps-Scripts/blob/main/docs/UiPathPack.md)

Output of the will be stored in \package

Uploading to Github

Change this to send artifact to S3 like we do with our other pipelines.

Open image-20240408-143620.png
image-20240408-143620.png
 

 

Deploy UIPath Package
Download Artifact

Use the UiPathDeploy.ps1 script to push the \package to Orchestrator Project

[UiPath-DevOps-Scripts/docs/UiPathDeploy.md at main · UiPath-Services/UiPath-DevOps-Scripts (github.com)](https://github.com/UiPath-Services/UiPath-DevOps-Scripts/blob/main/docs/UiPathDeploy.md)

Open image-20240408-143711.png
image-20240408-143711.png
 

 

Verify
 

Can we add some sort of verify step here to ensure Package was deployed successfully?

What about Automated Tests?

Future State
Potentially look into way to connect SNOW change controls with Production deployments to ensure a change control exists prior to merger or deployment.

How can we automate QA testing?

 

Resources
 

[CI/CD Pipelines for UiPath Process Deployment Series: GitHub Actions (youtube.com)](https://www.youtube.com/watch?v=musSykryH7U)

[UiPath Beginners Course 2023 - How to Get Started - YouTube](https://www.youtube.com/watch?v=A8cLOfItmv8)

UiPath RPA Developer Best Practices | Workflow Type in UiPath | RPA Projects | #shorts ([YouTube](http://youtube.com/) )

[UiPath DevOps Scripts - RPA Component | UiPath Marketplace | Overview](https://marketplace.uipath.com/listings/uipath-devops-scripts4936)

[UiPath Documentation Portal](https://docs.uipath.com/)

[Automation Ops - Executing Tasks (uipath.com)](https://docs.uipath.com/automation-ops/automation-cloud/latest/user-guide/executing-tasks-cli)

RPA Security

Owned by [Steven Luengo](https://bankunited.atlassian.net/wiki/people/5f454d86d0884f0049e91da0?ref=confluence&src=profilecard)

Last updated: [Apr 11, 2024](https://bankunited.atlassian.net/wiki/pages/diffpagesbyversion.action?pageId=3192291339&selectedPageVersions=2&selectedPageVersions=3)3 min read
2 people viewed
[Authentication & Authorization](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Authentication-&-Authorization)
[Secrets Management & Encryption](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Secrets-Management-&-Encryption)
[Compliance & Data Security](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Compliance-&-Data-Security)
[Auditing & Logging](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Auditing-&-Logging)
[Patching](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Patching)
[Application Security](https://bankunited.atlassian.net/wiki/spaces/WE/pages/3192291339/RPA+Security#Application-Security)
 

 

Authentication & Authorization
Secured Access to UIPath Orchestrator: Monitor, record, and isolate admin activity when admin users are logged into the console. Are these controlled via roles? Is this controlled in Azure/AD?

How is access controlled today to the orchestrator? Do you have RBAC in place?

Admin --> admin AD group

Developer --> developer AD group

Etc.

Current Status: NEEDS IMPROVEMENT 

Follow up with AIM, may have been re-prioritized (resourcing issue).

Right now, users are being provisioned via Orchestrator (local creds)

Groups are in place, but we need to re-evaluate role permissions and scope.

Roles:

 Prod Tenant ORG Admin: Jesse Revueltas, Admin-Service-Account

 UAT Tenant ORG Admin: Andres Quesada, Randy, Admin-Service-Account

Developer: Contractors are here 

Application Support: Contractors are here

Robot: For robot accounts

IAM: BankUnited IAM team

Is access granted automatically via SailPoint/SAML?

Current Status: VALIDATE/TEST

Is two-factor authentication enabled to this admin accounts?

Current Status: NEEDS IMPROVEMENT

This is not currently implemented.

 

Use unique Identifiers: Bots require their own unique identities/creds to access IT resources. To ensure nonrepudiation and separation of users.

Does each bot user have its own login credentials?

Current Status: NEEDS IMPROVEMENT

Bots are containers where we run automations. Bots are tied to Workstation (machine) and LDAP Identity.

Each Bot LDAP Identity has its own login credentials, but multiple automations (jobs/processes) can run from a single bot.

image-20240411-195920.png
  

We can improve this by migrating from classic folders to the new folder type. Then, we can segment bots by FOLDERS tied to DEPARTMENTS.

Each department folder will have its own set of bots tied to processes for that particular department.

This will be improved too when we move to Ec2 instances. Then each LDAP Identity will be tied to a particular department perhaps.

How are bot users currently provisioned?

Bot Identities in Classic Folders (Default) are tied to Machine/User

What is the naming convention for these accounts? 

Current Status: NEEDS IMPROVEMENT

Naming convention isn’t clear. Currently set as: BOT_ENVIRONMENT_NUMBER (underscore for legibility, actual bot names do not have them). However, bot names don’t always match LDAP User account.

We should agree to a better naming convention that will make it easy to know what process that bot is tied to. Perhaps something like BOT_ENVRONMENT_DEPARTMENT_PROCESS

Is it easy to track what bot is tied to which flow?

Current Status: GOOD

Yes, this is seen in Orchestrator. Source tells you what triggered the process.

Triggers are seen in the console and are usually cron jobs or queues. 

image-20240411-201713.png
 

image-20240411-201859.png
 

Can one bot interact with multiple automation or is it a 1:1 relationship?

Current Status: GOOD

One bot can interact with multiple automations (processes) because humans interact with multiple applications to do their jobs, so they link automations to bots that tie together common processes.

 

Least Privilege: Ensure the bot's access should be limited to just what applications and databases it needs for its job and nothing more.

How are bots currently limiting access? Is this done via AD groups? 

Do the AD groups limit access to what Apps and databases they can touch?

 

Secrets Management & Encryption
Remove Hard coded secrets: All hard-coded passwords/creds within scripts should be replaced with API call to a secret store.

How are secrets passed to scripts today?

Credential Store: Utilize a Centralized Credential store to be able to easily do things like password rotation, ensure uniqueness, and complexity.

Are you using a centralized secret store? If so, what?

How is access granted/controlled to these secrets?

Encryption: Is sensitive data encrypted and is access controlled? 

What algorithms are being used?

Where are encryption keys stored?

 

Compliance & Data Security
PII Data:

Are any flows dealing with PII? How are we masking PII data?

SmartCrypt? What is the responsibility of our team on this front?

Compliance Requirements

Are any flows we currently have handling data that has specific regulatory compliance requirements like PCI DSS? Are these identified/documented?

 

 

Auditing & Logging
Are all bot activates logged? If so, is it going to a centralized system?

What is the current retention policy on these logs?

Is there any log monitoring in place both for reliability and security issues?

 

Patching
As we move into Ec2 instances, we need to understand how these instances will be patched. SCCM? Tanium?

 

Application Security
Is any security code review done on scripts/flows today?
